stdlib/public/ClientRetainRelease/RetainRelease.s

```diff
@@ -0,0 +1,383 @@
+// stores performed on this thread prior to calling release.
+  casl  x16, x17, [x1]
+
+// The previous value of the refcount field is now in x16. We succeeded if that
+// value is the same as the old value we had before. If we failed, retry.
+  cmp   x2, x16
+  b.ne  Lrelease_retry
+#elif USE_LDX_STX
+// Try to store the updated value.
+  stlxr  w16, x17, [x1]
+
+// On failure, retry.
+  cbnz  w16, Lrelease_retry
+#else
+#error Either USE_CAS or USE_LDX_STX must be set.
+#endif
+
+// On success, return.
+Lrelease_ret:
+  ret
+
+Lslowpath_release:
+  CONDITIONAL USE_LDX_STX, \
+    clrex
+  CALL_SLOWPATH _swift_release_preservemost
+
+.alt_entry _bridgeObjectReleaseClientObjC
+_bridgeObjectReleaseClientObjC:
+  CONDITIONAL PTRAUTH, \
+    pacibsp
+  stp   x0, x9, [sp, #-0x50]!
+  stp   x10, x11, [sp, #0x10]
+  stp   x12, x13, [sp, #0x20]
+  stp   x14, x15, [sp, #0x30]
+  stp   fp, lr, [sp, #0x40];
+  add   fp, sp, #0x40
+
+  // Clear the unused bits from the pointer
+  and   x0, x0, #0x0ffffffffffffff8
+  bl    _objc_release
+
+  ldp   fp, lr, [sp, #0x40]
+  ldp   x14, x15, [sp, #0x30]
+  ldp   x12, x13, [sp, #0x20]
+  ldp   x10, x11, [sp, #0x10]
+  ldp   x0, x9, [sp], #0x50
+LbridgeObjectReleaseObjCRet:
+  CONDITIONAL PTRAUTH, \
+    retab
+  CONDITIONAL !PTRAUTH, \
+    ret
+
+
+.private_extern _swift_bridgeObjectRetainClient
+#if SWIFT_OBJC_INTEROP
+_swift_bridgeObjectRetainClient:
+  tbz  x0, #63, LbridgeObjectRetainNotTagged
+  ret
+LbridgeObjectRetainNotTagged:
+  tbnz  x0, #62, _swift_bridgeObjectRetainClientObjC
+
+.alt_entry _swift_retainClient
+#else
+.set _swift_bridgeObjectRetainClient, _swift_retainClient
+#endif
+
+.private_extern _swift_retainClient
+_swift_retainClient:
+// RR of NULL or values with high bit set is a no-op.
+  cmp   x0, #0
+  b.le  Lretain_ret
+
+// Mask off spare bits that may have come in from bridgeObjectRetain. Keep the
+// original value in x0 since we have to return it.
+  and   x1, x0, 0xffffffffffffff8
+
+// We'll operate on the address of the refcount field, which is 8 bytes into
+// the object.
+  add   x1, x1, #8
+
+// Load the current value of the refcount field when using CAS.
+  CONDITIONAL USE_CAS, \
+    ldr   x16, [x1]
+
+Lretain_retry:
+// Get the slow path mask and see if the refcount field has any of those bits
+// set.
+  adrp  x17, LretainRelease_slowpath_mask@PAGE
+  ldr   x17, [x17, LretainRelease_slowpath_mask@PAGEOFF]
+
+// Load-exclusive of the current value in the refcount field when using LLSC.
+// stxr does not update x16 like cas does, so this load must be inside the loop.
+// ldxr/stxr are not guaranteed to make forward progress if there are memory
+// accesses between them, so we need to do this after getting the mask above.
+  CONDITIONAL USE_LDX_STX, \
+    ldxr x16, [x1]
+
+  tst   x16, x17
+
+  b.ne  Lslowpath_retain
+
+// Compute a refcount field with the strong refcount incremented.
+// TODO: overflow checking
```

Comment:
Noting the presence of this TODO

Reply:
Good catch, thank you. Fixed now, and added a test that checks for the appropriate fatal error message.